src/Storages/ExternalStream/Kafka/KafkaSource.cpp

```diff
@@ -337,4 +348,72 @@ void KafkaSource::calculateColumnPositions()
         physical_header.insert({physical_column.type->createColumn(), physical_column.type, physical_column.name});
     }
 }
+
+
+/// It basically initiate a checkpoint
+/// Since the checkpoint method is called in a different thread (CheckpointCoordinator)
+/// We nee make sure it is thread safe
+void KafkaSource::checkpoint(CheckpointContextPtr ckpt_ctx_)
+{
+    /// We assume the previous ckpt is already done
+    /// Use std::atomic<std::shared_ptr<CheckpointContext>>
```

Comment:
This interface is called by another thread (cyclically called in CheckpointCoordinator::triggerCheckpoint), which notifies the source to generate a checkpoint barrier block, which is used to propagate to each processor downstream of the pipeline and do checkpoint (record its own status)

So here use an atomic vairable to save this ckpt request (i.e. `CheckpointContextPtr `) , it will be checked in `KafkaSource::generate()` at follow code:
```c++
Chunk KafkaSource::generate()
{
    ...
    if (auto * current = ckpt_ctx.exchange(nullptr, std::memory_order_relaxed); current)
        return doCheckpoint(CheckpointContextPtr{current});
    ...
}
```

For each processor, we need to do the following two things:
1) If received checkpoint request in current processor, do checkpoint
2) Output a checkpoint barrier chunk to downstream (continue to do `1) 2)` in next processor)

Reply:
In old clang++ version, the usage `std::atomic<std::shared_ptr<...>>` isn't supported, so we use a pointer here instead